cli/provider_cache.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	// By using `TF_CLI_CONFIG_FILE` we force terraform to use our auto-generated cli configuration file.
+	// https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_cli_config_file
+	envs[terraform.EnvNameTFCLIConfigFile] = cliConfigFile
+	// Clear this `TF_PLUGIN_CACHE_DIR` value since we are using our own caching mechanism.
+	// https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_plugin_cache_dir
+	envs[terraform.EnvNameTFPluginCacheDir] = ""
+
+	return envs
+}
+
+// createLocalCLIConfig creates a local CLI configuration that merges the default/user configuration with our Private Registry configuration.
+// We don't want to use Terraform's `plugin_cache_dir` feature because the cache is populated by our built-in private registry, and to make sure that no Terraform process ever overwrites the global cache, we clear this value.
+// In order to force Terraform to queries our registry instead of the original one, we use the section below.
+// https://github.com/hashicorp/terraform/issues/28309 (officially undocumented)
+//
+//	host "registry.terraform.io" {
+//		services = {
+//			"providers.v1" = "http://localhost:5758/v1/providers/registry.terraform.io/",
+//		}
+//	}
+//
+// In order to force Terraform to create symlinks from the provider cache instead of downloading large binary files, we use the section below.
+// https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation
+//
+//	provider_installation {
+//		filesystem_mirror {
+//			path    = "/path/to/the/provider/cache"
+//			include = ["example.com/*/*"]
+//		}
+//		direct {
+//			exclude = ["example.com/*/*"]
+//		}
+//	}
+//
+// This func doesn't change the default CLI config file, only creates a new one at the given path `cliConfigFile`. Ultimately, we can assign this path to `TF_CLI_CONFIG_FILE`.
+func createLocalCLIConfig(opts *options.TerragruntOptions, cliConfigFile string, registryProviderURL *url.URL) error {
+	cfg, err := terraform.LoadConfig()
+	if err != nil {
+		return err
+	}
+	cfg.PluginCacheDir = ""
+
+	providerInstallationIncludes := make([]string, len(opts.ProviderCacheRegistryNames))
+
+	for i, registryName := range opts.ProviderCacheRegistryNames {
+		cfg.AddHost(registryName, map[string]any{
+			"providers.v1": fmt.Sprintf("%s/%s/", registryProviderURL, registryName),
+		})
+
+		providerInstallationIncludes[i] = fmt.Sprintf("%s/*/*", registryName)
+	}
+
+	cfg.AddProviderInstallation(
+		terraform.NewProviderInstallationFilesystemMirror(opts.ProviderCacheDir, providerInstallationIncludes, nil),
+		terraform.NewProviderInstallationDirect(providerInstallationIncludes, nil),
+	)
+
+	if cfgDir := filepath.Dir(cliConfigFile); !util.FileExists(cfgDir) {
+		if err := os.MkdirAll(cfgDir, os.ModePerm); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	if err := cfg.Save(cliConfigFile); err != nil {
+		return err
+	}
+	return nil
```

Comment:
err check seems to be redundant 
```suggestion
	return cfg.Save(cliConfigFile)
```

Reply:
You are right, simplified the code.